lib/pbench/cli/agent/commands/tools/kill.py

```diff
@@ -0,0 +1,249 @@
+  2. Find the recorded pids for the Redis Server, Tool Data Sink, and local
+     Tool Meister in their respective pid files, and stop those processes
+     from running
+  3. Determine all the remote hosts used for that run
+  4. For each remote host:
+     a. `ssh` to that remote host
+     b. Stop the Tool Meister running on that host
+
+The pbench-tool-meister-start uses a UUID in the command line of all the
+remote Tool Meister processes. Any Tool Meister process with that UUID in its
+command line string will be stopped via `kill -KILL`, along with all of its
+child processes.
+"""
+
+from collections import defaultdict
+import pathlib
+import shlex
+from typing import Callable, List, Tuple
+
+import click
+import psutil
+
+from pbench.agent.base import BaseCommand
+from pbench.agent.tool_group import ToolGroup
+from pbench.agent.utils import LocalRemoteHost, TemplateSsh
+from pbench.cli.agent import CliContext, pass_cli_context
+from pbench.cli.agent.options import common_options
+
+
+def kill_family(proc: psutil.Process):
+    """Kill a parent process and all its children."""
+    try:
+        # Get the list of children of the parent before killing it.
+        children = list(proc.children())
+    except psutil.NoSuchProcess:
+        return
+    try:
+        proc.kill()
+    except psutil.NoSuchProcess:
+        pass
+    for child in children:
+        try:
+            child.kill()
+        except psutil.NoSuchProcess:
+            pass
+
+
+class PidSource:
+    """For a given PID file name keep track of discovered PIDs and UUIDs as
+    Tool Meister directories are `load()`ed.  The `killem()` method is invoked
+    by the caller at their discretion.
+
+    The `killem()` method clears out all accumlated data.
+    """
+
+    def __init__(self, file_name: str, display_name: str):
+        self.file_name = file_name
+        self.display_name = display_name
+        self.pids_by_uuid = {}
+        self.uuid_to_tmdir = {}
+
+    def load(self, tm_dir: pathlib.Path, uuid: str) -> bool:
+        """Load a PID from the given directory associated with the given UUID.
+
+        Records the loaded PID if it has a live process associated with it and
+        returns True, otherwise returns False.
+        """
+        try:
+            pid = (tm_dir / self.file_name).read_text()
+        except FileNotFoundError:
+            return False
+        try:
+            self.pids_by_uuid[uuid] = psutil.Process(pid)
+        except psutil.NoSuchProcess:
+            return False
+        self.uuid_to_tmdir[uuid] = tm_dir
+        return True
+
+    def killem(self, echo: Callable[[str], None]) -> None:
+        """Kill all PIDs found, and their children."""
+        if not self.pids_by_uuid:
+            return
+        echo(f"Killing {self.display_name} PIDs ...")
+        pids_by_uuid, self.pids_by_uuid = self.pids_by_uuid, {}
+        uuid_to_tmdir, self.uuid_to_tmdir = self.uuid_to_tmdir, {}
```

Comment:
This "fetch and clear" idiom is cute but a bit obscure: I'd rather see these broken up, I think!

Reply:
This is such a common python code sequence that breaking it up would be an afront to the language.  =)  I certainly can add a comment.

Reply:
Sure, the syntax is common, and the semantics here are clever but I think unnecessarily obscured by the common syntax. But whatever: it's a small section, and it works.